### ISSUE TITLE
PyTango raises an unexpected exception

### ISSUE THREAD

**Issue by cmft** [tango-controls/pytango#89]
PyTango raises an unexpected exception if you try to write an attribute when 
device is not running.  This behaviour is different between PyTango9 and older
versions.

If you try to read the attribute PyTango raises a DevError exception as it is expected 
but if you try to write it raises TypeError: unsupported data_format

I'm using PyTango 9.2.0 (From TAG) installed with package manager from https://build.opensuse.org/project/show/home:cmft

Here the trace:

```python
import PyTango
d = PyTango.DeviceProxy('sys/tg_test/1')

## Read

d.read_attribute('short_scalar')
---------------------------------------------------------------------------
ConnectionFailed                          Traceback (most recent call last)
<ipython-input-4-375a92ead2c8> in <module>()
----> 1 d.read_attribute('short_scalar')

/usr/lib64/python2.7/site-packages/tango/green.pyc in greener(self, *args, **kwargs)
    264 
    265         # submit the method
--> 266         ret = submitter(fn, self, *args, **kwargs)
    267 
    268         # return the proper result

/usr/lib64/python2.7/site-packages/tango/green.pyc in submit(self, fn, *args, **kwargs)
     98 class SynchExecutor(object):
     99     def submit(self, fn, *args, **kwargs):
--> 100         return fn(*args, **kwargs)
    101 
    102 __synch_executor = SynchExecutor()

/usr/lib64/python2.7/site-packages/tango/device_proxy.pyc in __DeviceProxy__read_attribute(self, value, extract_as)
    328 
    329 def __DeviceProxy__read_attribute(self, value, extract_as=ExtractAs.Numpy):
--> 330     return __check_read_attribute(self._read_attribute(value, extract_as))
    331 
    332 #def __DeviceProxy__read_attribute(self, value, extract_as=ExtractAs.Numpy,

ConnectionFailed: DevFailed[
DevError[
    desc = Device sys/tg_test/1 is not exported (hint: try starting the device server)
  origin = DeviceProxy::get_corba_name()
  reason = API_DeviceNotExported
severity = ERR]

DevError[
    desc = Failed to read_attribute on device sys/tg_test/1, attribute short_scalar
  origin = DeviceProxy::read_attribute()
  reason = API_AttributeFailed
severity = ERR

## Write

d.write_attribute('short_scalar', 1)
<ipython-input-5-74477859a0f7> in <module>()
----> 1 d.write_attribute('short_scalar', 1)

/usr/lib64/python2.7/site-packages/tango/green.pyc in greener(self, *args, **kwargs)
    264 
    265         # submit the method
--> 266         ret = submitter(fn, self, *args, **kwargs)
    267 
    268         # return the proper result

/usr/lib64/python2.7/site-packages/tango/green.pyc in submit(self, fn, *args, **kwargs)
     98 class SynchExecutor(object):
     99     def submit(self, fn, *args, **kwargs):
--> 100         return fn(*args, **kwargs)
    101 
    102 __synch_executor = SynchExecutor()

/usr/lib64/python2.7/site-packages/tango/device_proxy.pyc in __DeviceProxy__write_attribute(self, *args, **kwargs)
   1226 
   1227 def __DeviceProxy__write_attribute(self, *args, **kwargs):
-> 1228     return self._write_attribute(*args, **kwargs)
   1229 
   1230 def __DeviceProxy__write_attributes(self, *args, **kwargs):

TypeError: unsupported data_format.
```

**Comment by AntoineDupre** [comment]
Thank you for the report,

This ErrorType is raised in the Boost layer. 

_device_proxy.cpp:_
```cpp
    static inline void write_attribute(Tango::DeviceProxy& self, const string & attr_name, bopy::object py_value)
    {
        Tango::DeviceAttribute dev_attr;
        PyDeviceAttribute::reset(dev_attr, attr_name, self, py_value);
        {
            AutoPythonAllowThreads guard;
            self.write_attribute(dev_attr);
        }
    }
```
Before writing an attribute, we try to get the attribute configuration with```PyDeviceAttribute::reset``` 

_device_attribute.cpp:_
```cpp
  reset(Tango::DeviceAttribute & self, const std::string &attr_name,
          Tango::DeviceProxy &dev_proxy, bopy::object py_value)
    {
        self.set_name(attr_name.c_str());
        Tango::AttributeInfoEx attr_info;
        {
            AutoPythonAllowThreads guard;
            try
            {
                attr_info = dev_proxy.get_attribute_config(attr_name);
            }
            /catch(...)
            {}
        }
        reset_values(self, attr_info.data_type, attr_info.data_format, py_value);
    }

```

The```get_attribute_config```method is inside a```try/catch```section. If the device server is not running, the```ConnectionFailed```exception raised by this method is swallowed, the```attr_info.data_format```is not initialized and the```reset_values```method raises a```TypeError```exception.

_device_attribute.cpp:_
```cpp
    reset_values(Tango::DeviceAttribute & self, int data_type,
                 Tango::AttrDataFormat data_format, bopy::object py_value)
    {
        bool isImage = false;
        switch(data_format)
        {
            case Tango::SCALAR:
                 ///...
            case Tango::IMAGE:
                ///...
            case Tango::SPECTRUM:
                ///...
            default:
                raise_(PyExc_TypeError, "unsupported data_format.");
        }
    }

```

This```try/catch```section has been added in commit [ee7eb27eff](https://github.com/tango-controls/pytango/commit/ee7eb27eff383f82061d0f59335a419193968067) to avoid a segfault with Group object (see issue [#637](https://sourceforge.net/p/tango-cs/bugs/637/) ).

@tiagocoutinho Do you know if this issue637 was in fact related to cppTango and if it has been fixed in the C++ layer ?

**Comment by jairomoldes** [comment]
This problem seems to be solved in 9.2.4
A build of pytango 9.2.4 over cpp tango 9.2.5 raises a DevFailed exception as expected.
I propose to close this issue

**Comment by jairomoldes** [comment]
Problem seems solved.
